Added documentation for --ignore-weak-subjectivity-period-enabled

lucassaldanha opened the pull request:
Documentation for https://github.com/Consensys/teku/pull/7698

alexandratran commented:
@lucassaldanha
Made some edits that require review:
- Moved the new content to the [weak subjectivity concepts page](https://doc-teku-cnzuygc06-infura-web.vercel.app/development/concepts/weak-subjectivity)
- Link to weak subjectivity at the top of the [checkpoint sync page](https://doc-teku-cnzuygc06-infura-web.vercel.app/development/get-started/checkpoint-start)
- Reorder reference so [`checkpoint-sync-url`](https://doc-teku-cnzuygc06-infura-web.vercel.app/development/reference/cli#checkpoint-sync-url) is in alphabetical order

Questions:
1. What's the difference between `checkpoint-sync-url` and `initial-state`?
2. Do we recommend checkpoint sync for most users?
3. Do you need to specify `ignore-weak-subjectivity-period-enabled=true` if you don't use a checkpoint state?

lucassaldanha commented:
> What's the difference between `checkpoint-sync-url` and `initial-state`?

They work in a similar way (behind the scenes, we set the initial-state value based on the `checkpoint-sync-url`.

The main reason for using it is to be aligned with the options that other clients are using. Checkpoint sync is a common term that users often refer to when syncing their nodes.

One big difference is that `initial-state` also accepts a file, not just a URL.

> Do we recommend checkpoint sync for most users?

Totally. This is one of the reasons why we changed the default syncing behaviour to forbid users to sync from old states (including genesis).

> Do you need to specify `ignore-weak-subjectivity-period-enabled=true` if you don't use a checkpoint state?

Most likely yes, unless you are on some private network just starting. This is how it plays:
1. Start a node w/o checkpoint sync URL on network X
2. Teku will check the genesis state for network X
3. If the state is too far behind, Teku will stop and give an error. Unless `ignore-weak-subjectivity-period-enabled=true`.

If you are starting your node close to the genesis time of the network, then it is possible that we are still within the weak subjectivity period, so step 3 won't fail.

However, for mainnet and all official current testnets, this won't happen because their genesis is already outside of the weak subjectivity period.

lucassaldanha commented on docs/concepts/weak-subjectivity.md:
```diff
@@ -6,31 +6,62 @@ sidebar_position: 7
-We recommend using `--initial-state`. It provides the same security benefits as `--ws-checkpoint`, but syncs faster. The only exception is when syncing an archive node, in which case, use `--ws-checkpoint`.
+Teku provides three methods to safely sync a node that is new to the network or has been offline for an extended period.
 
-:::tip
+1. Use [`--checkpoint-sync-url`](../reference/cli/index.md#checkpoint-sync-url) to supply a URL of a
+    checkpoint state endpoint from which to sync.
+2. Use [`--initial-state`](../reference/cli/index.md#initial-state) to supply an SSZ encoded state
+    file from which to sync.
+3. Use [`--ws-checkpoint`](../reference/cli/index.md#ws-checkpoint) to supply a weak subjectivity
+    checkpoint by which a node can securely validate its view of the current state.
 
-Use the [`/eth/v2/debug/beacon/states/<state_id>`](https://consensys.github.io/teku/#operation/getEthV2DebugBeaconStatesWithState_id) API on an updated node to download a recent finalized state as an SSZ encoded state file.
+We recommend using `--checkpoint-sync-url` or `--initial-state`.
+They provide the same security benefits as `--ws-checkpoint`, but sync faster.
+The only exception is when syncing an archive node, in which case, use `--ws-checkpoint`.
```
I think we might need to revisit this recommendation. For mainnet, we should only recommend people use `checkpoint-sync-url`.

lucassaldanha commented on docs/get-started/checkpoint-start.md:
```diff
@@ -6,16 +6,19 @@ sidebar_position: 4
 # Start Teku from a recent state
 
-To get Teku up and running in only a few minutes, you can start Teku from a recent finalized checkpoint state rather than syncing from genesis.
+To get Teku up and running in only a few minutes, you can start Teku from a recent finalized
```
maybe we change `you can` for `you should` to emphasise `checkpoint-sync-url` is better?

alexandratran commented:
@lucassaldanha Thanks for the thorough response. I've addressed your feedback, my only outstanding question is whether we need to update the examples on the following pages to use `checkpoint-sync-url`?

- https://docs.teku.consensys.net/get-started/start-teku
- https://docs.teku.consensys.net/get-started/connect/mainnet
- https://docs.teku.consensys.net/get-started/connect/testnet
- (Besu docs) https://besu.hyperledger.org/public-networks/tutorials/besu-teku-mainnet
- (Besu docs) https://besu.hyperledger.org/public-networks/tutorials/besu-teku-testnet

lucassaldanha commented:
> @lucassaldanha Thanks for the thorough response. I've addressed your feedback, my only outstanding question is whether we need to update the examples on the following pages to use `checkpoint-sync-url`?
> 
> * https://docs.teku.consensys.net/get-started/start-teku
> * https://docs.teku.consensys.net/get-started/connect/mainnet
> * https://docs.teku.consensys.net/get-started/connect/testnet
> * (Besu docs) https://besu.hyperledger.org/public-networks/tutorials/besu-teku-mainnet
> * (Besu docs) https://besu.hyperledger.org/public-networks/tutorials/besu-teku-testnet

Hi Alexandra, yes I think it would be great to update the example on these pages to use `--checkpoint-sync-url`. The only thing is what checkpoint sync URL we use in the examples. Basically, we do not want to make the decision for people on what "source of truth" they should trust. Even though in practice, using an endpoint like Ethstaker's is the sensible thing to do.

So the options are:
1. We chose a trustworthy checkpoint sync provider and use it in the examples (e.g. Ethstaker's).
2. We use a generic endpoint as a placeholder (e.g. `http://your-chosen-checkpoint-sync-url.com`).

Option 1 is more friendly to users, but it is more opinionated on what endpoint to use. @benjaminion do you have some thoughts on this?

alexandratran commented:
@lucassaldanha I used the Ethstaker endpoint for the "Start Teku" page because those are clearly labelled as example commands. For the connect to Mainnet and testnet pages, I added placeholder text and linked to the reference. Also raised this issue to update the Besu pages: https://github.com/hyperledger/besu-docs/issues/1451

lucassaldanha commented on docs/get-started/connect/mainnet.md:
```diff
@@ -82,10 +82,16 @@ teku \
     --metrics-enabled=true                       \
-    --rest-api-enabled=true
+    --rest-api-enabled=true                      \
+    --checkpoint-sync-url=<checkpoint state URL>
```
Let's use `<checkpoint sync URL>` for the placeholder to avoid any confusion with users. That applies to the other examples as well.
